metaflow/plugins/argo/argo_events_decorator.py

```diff
@@ -0,0 +1,154 @@
+from metaflow.decorators import FlowDecorator
+from metaflow.exception import MetaflowException
+from metaflow.util import is_stringish
+
+# TODO: Support dynamic parameter mapping through a context object that exposes
+#       flow name and user name similar to parameter context
+
+
+# TODO: This decorator interface can be a top-level decorator to support similar
+#       implementations for Netflix Maestro, AWS Step Functions, Apache Airflow...
+class ArgoEventsDecorator(FlowDecorator):
+    name = "trigger"
+    defaults = {
+        "event": None,
+        "events": [],
+        "options": {},  # TODO: introduce support for options
+    }
+
+    def flow_init(
+        self, flow, graph, environment, flow_datastore, metadata, logger, echo, options
+    ):
+        self.triggers = []
+        if sum(map(bool, (self.attributes["event"], self.attributes["events"]))) > 1:
+            raise MetaflowException(
+                "Specify only one of *event* or *events* "
+                "attributes in *@trigger* decorator."
+            )
+        elif self.attributes["event"]:
+            # event attribute supports the following formats -
+            #     1. event='table.prod_db.members'
+            #     2. event={'name': 'table.prod_db.members',
+            #               'parameters': {'alpha': 'member_weight'}}
+            if is_stringish(self.attributes["event"]):
+                self.triggers.append({"name": str(self.attributes["event"])})
+            elif isinstance(self.attributes["event"], dict):
+                if "name" not in dict(self.attributes["event"]):
+                    raise MetaflowException(
+                        "The *event* attribute for *@trigger* is missing the "
```

Comment:
Should we pull the common parse/validation logic out?  Duplicated across "event" and "events" right now.

Reply:
I tried doing that. Given the subtle changes in the error messages, it would have been quite a bit more complicated than simply repeating it. Definitely scope for improvement though. We can do another pass post merge.